vioscsi/vioscsi.c

```diff
@@ -438,60 +438,189 @@ VioScsiFindAdapter(
+    ConfigInfo->NumberOfPhysicalBreaks = SCSI_MINIMUM_PHYSICAL_BREAKS; // 16
+
+    if (!adaptExt->dump_mode) {
         adaptExt->max_physical_breaks = PHYS_SEGMENTS;
+
+        /* Check and enable indirect descriptor support. If available,
+         * reevaluate various limits to take full advantage of indirection.
+         *
+         * With indirect descriptors, we can use a larger amount of physical
+         * segments (aka physical breaks) in a single transfer. Each indirect
+         * descriptor based request will only occupy a single entry on the
+         * virtio ring. This entry is only a pointer to a separate indirect
+         * descriptor chain containing the actual request.
+         *
+         * IMPORTANT NOTE: To get good large IO performance we need to bend
+         * the virtio rules a little.
+         *
+         * Spec reference:
+         *   https://docs.oasis-open.org/virtio/virtio/v1.1/virtio-v1.1.html
+         *   2.6.5.3.1 Driver Requirements: Indirect Descriptors
+         *   A driver MUST NOT create a descriptor chain longer than the Queue
+         *   Size of the device.
+         *
+         * Example using 1 MByte IO:
+         * Assuming non-contiguous memory, a single 1MB IO will consist of
+         * 256x 4096 byte buffers (Windows PAGE_SIZE), which would translate
+         * to a sgList of 256 elements.
+         *
+         * If we stuck with the letter of the specifications at time of
+         * writing, we'd need a virtio ring of at least 258 in size, even
+         * though the actual request is stored as a single entry on the
+         * ring. Rings that large are uncommon as support for larger and/or
+         * easily configurable ring sizes wasn't available in QEMU until
+         * 5.0+. Even then, its a guest visible parameter change, which may
+         * not be trivial for consumers to implement.
+         *
+         * Given that there is no way for a guest to actually tell if their
+         * underlying device will support these super sized SG lists or not,
+         * we'll try to be as sane about this as possible, which includes
+         * a "break glass" configuration override knob in the Windows registry.
+         *
+         * The maximum transfer length can be adjusted by creating and editing
+         * the following Registry key, which the code below will read and apply.
+         * [HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\vioscsi\Parameters\Device]
+         * "PhysicalBreaks"={dword value here}
+         *
+         * TODO: Update this logic to incorporate indirect descriptor queue
+         * size limits if/when it is fully ratified into the virtio spec.
+         * This will allow the guest and host to negotiate on this specific
+         * limit. Reference:
+         *   https://github.com/oasis-tcs/virtio-spec/issues/122
+         */
+        adaptExt->indirect = CHECKBIT(adaptExt->features, VIRTIO_RING_F_INDIRECT_DESC);
         if (adaptExt->indirect) {
```

Comment:
Btw, does any real reason even exist for this check? I mean, the fact that we have two default values disgusts me. For one it's not like `32` ("`PHY_SEGMENTS`") is always safe enough.

For example, for devices with max_(hw_)sector being 240 (USB 2.0 drive), supposedly it can only deal with NOPB being 16+1.

Besides, it would look like that we are suggesting indirect_desc should be set to off when we bump into problems, but the real problem might be *just* the fact that NOPB is too high (i.e. we need to set a clamp).

Let alone putting the registry reading inside a condition.

Reply:
Keep in mind my comments are from an the point of view of an outside contributor, but IMHO yes there is a point for this check, as the mere presence of adapter supported indirect descriptors means that we dont have to be as careful about filling up the actual virtio ring itself. That seems to be the original intention of the code

As for USB drive limits -> take this comment with a grain of salt, but do we care? I'd imagine most of the stuff that users would use vioscsi on is not a USB drive, but some sort of more capable storage device :) 

Anyhow, you bring up a good point, let me see if I can clean this up a bit more